gallery/diff3d/ceed-diff3DBuild.h

```diff
@@ -0,0 +1,57 @@
+// Copyright (c) 2017-2018, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory. LLNL-CODE-734707.
+// All Rights reserved. See files LICENSE and NOTICE for details.
+//
+// This file is part of CEED, a collection of benchmarks, miniapps, software
+// libraries and APIs for efficient high-order finite element and spectral
+// element discretizations for exascale applications. For more information and
+// source code availability see http://github.com/ceed.
+//
+// The CEED research is supported by the Exascale Computing Project 17-SC-20-SC,
+// a collaborative effort of two U.S. Department of Energy organizations (Office
+// of Science and the National Nuclear Security Administration) responsible for
+// the planning and preparation of a capable exascale ecosystem, including
+// software, applications, hardware, advanced system engineering and early
+// testbed platforms, in support of the nation's exascale computing imperative.
+
+/**
+  @brief Ceed QFunction for building the geometric data for the 3D diffusion
+           operator
+**/
+CEED_QFUNCTION(diff3DBuild)(void *ctx, const CeedInt Q,
+                            const CeedScalar *const *in, CeedScalar *const *out) {
+  // At every quadrature point, compute qw/det(J).adj(J).adj(J)^T and store
+  // the symmetric part of the result.
+
+  // in[0] is Jacobians with shape [3, nc=3, Q]
+  // in[1] is quadrature weights, size (Q)
+  const CeedScalar *J = in[0], *qw = in[1];
+
+  // out[0] is qdata, size (Q)
+  CeedScalar *qd = out[0];
+
+  // Quadrature point loop
+  for (CeedInt i=0; i<Q; i++) {
+    // Compute the adjoint
+    CeedScalar A[3][3];
+    for (CeedInt j=0; j<3; j++)
+      for (CeedInt k=0; k<3; k++)
+//      A[k][j] = J[j+1][k+1]*J[j+2][k+2] - J[j+1][k+2]*J[j+2][k+1]
```

Comment:
Commented code is smelly code. Looks good otherwise.

Reply:
That's not intended as dead code but rather a simplified version of the code below for reference/readability. I can try to make that clearer? Without the VLAs, messy indices like `i+Q*((j+1)%3+3*((k+1)%3))` are somewhat slow to parse by humans, imho.

Reply:
Oh, gotcha.

Reply:
Added words to clarify